utils/contracts/governor.ts

```diff
@@ -0,0 +1,211 @@
+  if (governorType === 'bravo') {
+    // Get voting token and total supply
+    const govSlots = getBravoSlots(proposalId)
+    const rawVotingToken = await provider.getStorageAt(governor.address, govSlots.votingToken)
+    const votingToken = getAddress(`0x${rawVotingToken.slice(26)}`)
+    return erc20(votingToken)
+  }
+
+  return erc20(await governor.token())
+}
+
+export function getGovSlots(governorType: GovernorType, proposalId: BigNumberish) {
+  if (governorType === 'bravo') return getBravoSlots(proposalId)
+  return getOzSlots(proposalId)
+}
+
+export async function getProposalIds(
+  governorType: GovernorType,
+  address: string,
+  latestBlockNum: number
+): Promise<BigNumber[]> {
+  if (governorType === 'bravo') {
+    // Fetch all proposal IDs
+    const governor = governorBravo(address)
+    const proposalCreatedLogs = await governor.queryFilter(governor.filters.ProposalCreated(), 0, latestBlockNum)
+    const allProposalIds = proposalCreatedLogs.map((logs) => (logs.args as unknown as ProposalEvent).id!)
+
+    // Remove proposals from GovernorAlpha based on the initial GovernorBravo proposal ID
+    const initialProposalId = await governor.initialProposalId()
+    return allProposalIds.filter((id) => id.gt(initialProposalId))
+  }
+
+  const governor = governorOz(address)
+  const proposalCreatedLogs = await governor.queryFilter(governor.filters.ProposalCreated(), 0, latestBlockNum)
+  return proposalCreatedLogs.map((logs) => (logs.args as unknown as ProposalEvent).proposalId!)
+}
+
+export function getProposalId(proposal: ProposalEvent): BigNumber {
+  const id = proposal.id || proposal.proposalId
+  if (!id) throw new Error(`Proposal ID not found for proposal: ${JSON.stringify(proposal)}`)
+  return id
+}
+
+// Generate proposal ID, used when simulating new proposals.
+export async function generateProposalId(
+  governorType: GovernorType,
+  address: string,
+  // Below arg is only required for OZ governors.
+  {
+    targets,
+    values,
+    calldatas,
+    description,
+  }: { targets: string[]; values: BigNumberish[]; calldatas: string[]; description: string } = {
+    targets: [],
+    values: [],
+    calldatas: [],
+    description: '',
+  }
+): Promise<BigNumber> {
+  // Fetch proposal count from the contract and increment it by 1.
+  if (governorType === 'bravo') {
+    const count: BigNumber = await governorBravo(address).proposalCount()
+    return count.add(1)
+  }
+
+  // Compute proposal ID from the tx data
+  return BigNumber.from(
+    keccak256(
+      defaultAbiCoder.encode(
+        ['address[]', 'uint256[]', 'bytes[]', 'bytes32'],
+        [targets, values, calldatas, keccak256(toUtf8Bytes(description))]
+      )
+    )
+  )
+}
+
+// Returns the identifier of an operation containing a batch of transactions.
+// For OZ governors, predecessor is often zero and salt is often description hash.
+// This is only intended to be used with OZ governors.
+export function hashOperationBatch(
```

Comment:
nit: consider, feel free to ignore!

```suggestion
export function hashOperationBatchOZ(
```

Reply:
Resolved ✅